zcash_client_backend/src/scanning.rs

```diff
@@ -365,22 +365,57 @@ pub(crate) fn scan_block_with_runner<
             if m.sapling_commitment_tree_size == 0 {
                 None
             } else {
-                let block_note_count: u32 = block
+                let block_output_count: u32 = block
                     .vtx
                     .iter()
-                    .map(|tx| {
-                        u32::try_from(tx.outputs.len()).expect("output count cannot exceed a u32")
-                    })
-                    .sum();
-                Some(m.sapling_commitment_tree_size - block_note_count)
+                    .map(|tx| tx.outputs.len())
+                    .sum::<usize>()
+                    .try_into()
+                    .expect("Sapling output count cannot exceed a u32");
+                Some(m.sapling_commitment_tree_size - block_output_count)
             }
         })
         .or_else(|| prior_block_metadata.map(|m| m.sapling_tree_size()))
```

Comment:
While looking into the off-by-one bug, I realised that this code is backwards. If we have a `prior_block_metadata`, then (because `BlockMetadata` stores non-optional tree sizes, and we migrated the SQLite database to fill in Sapling data) we are guaranteed to have the correct value, so we should just use it directly instead of recomputing from the block's `chain_metadata` (which ends up performing a no-op computation, subtracting all the output counts and then re-adding them incrementally). That would amount to swapping the order of using `prior_block_metadata` and `block.chain_metadata`.

We should separately consider checking at the end of the scan that `sapling_commitment_tree_size` is equal to whatever is in `block.chain_metadata` (if present), as another arm of the consistency check (if we didn't have `prior_block_metadata` and thus weren't doing consistency checks, then this will be a check of the no-op-ness of the `block.chain_metadata` usage here).